caffe2/CMakeLists.txt

```diff
@@ -487,6 +489,21 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     ${TORCH_SRC_DIR}/csrc/jit/mobile/type_parser.cpp
     )
 
+  if (USE_LLVM)
+    message(STATUS "Looking for LLVM in ${USE_LLVM}")
+    find_package(LLVM QUIET PATHS ${USE_LLVM} NO_DEFAULT_PATH)
+
+    if (LLVM_FOUND)
+      message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION}")
+      message(STATUS "Using LLVMConfig.cmake in: ${LLVM_DIR}")
+
+      include_directories(${LLVM_INCLUDE_DIRS})
+      add_definitions(-DENABLE_LLVM ${LLVM_DEFINITIONS})
```

Comment:
Uhhh, as far as I can tell, `LLVM_DEFINITIONS` isn't used anywhere. Did you add a usage in a different PR?

Note that adding `-D` flags is an antipattern unless you can guarantee that `ENABLE_LLVM` only is in c++ files, as in general clients compiling against your library won't set this definition. But I see usage in headers... you should put this in one of our Config.h headers in that case

Reply:
And if you do put it in a Config.h, please prefix it with TORCH following our other definitions

Reply:
Would it be better to just conditionally include the llvm-related sources?  The only thing we really do with ENABLE_LLVM is to #ifdef out entire files.

Reply:
I'm not sure what you mean by "conditionally include"

Reply:
Oh, I mean in cmakelists, do something like 
if (USE_LLVM)
list(append TORCH_SRCS llvm_codegen.cpp ...)
endif()

Reply:
One important thing to note is that nothing internally hard depends on LLVM and LLVM is not exposed in any way to the user of libtorch. However, if libtorch was build with USE_LLVM, it gets an LLVM codegen for tensor expression engine (the codegen is registered via static constructors and is picked dynamically based on what is available). In that sense, I don't think it's important for whoever links with pytorch to get the value of `ENABLE_LLVM` used when pytorch was built.

Reply:
That said, I'll give it a shot to register LLVM as a usual dependency.

Reply:
@bertmaher So, I am perfectly happy to believe that things will "work out ok" the way things are setup today. What I am not convinced about, is when people who are not you start touching this code, they might start including the header from different places... some of which are publicly accessible... and then weird things start breaking and they don't know why. Unless you are going to add a lint rule enforcing whatever informal rule you've specified here ("nothing internally hard depends on LLVM and LLVM is not exposed in any way to the user of libtorch"), it's better to arrange for it to just be impossible to violate this internal rule, or make the consequences of violating it negligible (put it in a generated Config.h header).

Reply:
@ezyang I'm confused -- I didn't assert that everything will "work out ok" the way things are set up, and I wasn't suggesting we leave it alone.  I noticed that there are several other places in caffe2/CMakeLists.txt where we conditionally append to TORCH_SRCS based on other configuration variables, and I wondered if that was a better approach than using any #ifdefs at all.

Reply:
When I said "work out ok", I was referring specifically to this comment from @ZolotukhinM 

>  In that sense, I don't think it's important for whoever links with pytorch to get the value of ENABLE_LLVM used when pytorch was built.

Sorry about the confusing mention. 

> I noticed that there are several other places in caffe2/CMakeLists.txt where we conditionally append to TORCH_SRCS based on other configuration variables, and I wondered if that was a better approach than using any #ifdefs at all.

This seems better for now, at least.

Reply:
Ahh thanks for the clarification @ezyang :-).  That makes sense to me.

Reply:
> Would it be better to just conditionally include the llvm-related sources? The only thing we really do with ENABLE_LLVM is to #ifdef out entire files.

@bertmaher, we actually also use it for selecting a codegen for CPU in kernel.cpp.